chain/client/src/view_client.rs

```diff
@@ -758,6 +758,58 @@ impl Handler<GetStateChangesWithCauseInBlock> for ViewClientActor {
     }
 }
 
+/// Returns a hashmap where the key represents the ShardID and the value
+/// is the list of changes in a store with causes for a given block.
+impl Handler<GetStateChangesWithCauseInBlockForTrackedShards> for ViewClientActor {
+    type Result = Result<HashMap<ShardId, StateChangesView>, GetStateChangesError>;
+
+    #[perf]
+    fn handle(
+        &mut self,
+        msg: GetStateChangesWithCauseInBlockForTrackedShards,
+        _: &mut Self::Context,
+    ) -> Self::Result {
+        let state_changes_with_cause_in_block =
+            self.chain.store().get_state_changes_with_cause_in_block(&msg.block_hash)?;
+
+        let mut state_changes_with_cause_split_by_shard_id: HashMap<ShardId, StateChangesView> =
+            HashMap::new();
+        for state_change_with_cause in state_changes_with_cause_in_block {
+            let account_id = match &state_change_with_cause.value {
+                StateChangeValue::AccountUpdate { account_id, .. } => account_id.clone(),
+                StateChangeValue::AccountDeletion { account_id } => account_id.clone(),
+                StateChangeValue::AccessKeyUpdate { account_id, .. } => account_id.clone(),
+                StateChangeValue::AccessKeyDeletion { account_id, .. } => account_id.clone(),
+                StateChangeValue::DataUpdate { account_id, .. } => account_id.clone(),
+                StateChangeValue::DataDeletion { account_id, .. } => account_id.clone(),
+                StateChangeValue::ContractCodeUpdate { account_id, .. } => account_id.clone(),
+                StateChangeValue::ContractCodeDeletion { account_id } => account_id.clone(),
+            };
+            let shard_id = if let Ok(shard_id) =
+                self.runtime_adapter.account_id_to_shard_id(&account_id, &msg.epoch_id)
+            {
+                shard_id
+            } else {
+                return Err(GetStateChangesError::IOError {
+                    error_message: format!("Failed to get ShardID from AccountID {}", account_id),
```

Comment:
According to the `account_id_to_shard_id` implementation, the error could happen only when we cannot get shard layout (either due to some internal error or when the epoch never happened), so while it is more or less fair to abuse IOError, I would at least make the error message more helpful to both future developers and those who might potentially observe the error message. Thus, I suggest we format Err into the error message.